src/ReplicatedStorage/Client/InstanceComponents/Misc/CameraLook.lua

```diff
@@ -0,0 +1,164 @@
+local Players = game:GetService("Players")
+local RunService = game:GetService("RunService")
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+local TweenService = game:GetService("TweenService")
+local Trove = require(ReplicatedStorage.Packages.Trove)
+
+local CameraLook = {}
+CameraLook.__index = CameraLook
+
+local function calculateJointAngle(lookVector, sourcePosition, targetPosition)
+	return math.asin((sourcePosition - targetPosition):Dot(lookVector)/(sourcePosition - targetPosition).Magnitude)
+end
+
+local easingStyle = Enum.EasingStyle.Circular
+local easingDirection = Enum.EasingDirection.In
+local function dampenAngle(sourceAngle: number, goalAngle: number, angularSpeed: number, deltaTime: number)
+	local angleDiff = goalAngle%math.rad(360) - sourceAngle%math.rad(360)
+	if math.abs(angleDiff) > math.rad(180) then
+		local dividend = math.sign(angleDiff)*math.rad(180)
+		angleDiff = angleDiff%dividend - dividend
+	end
+	return sourceAngle + math.sign(angleDiff) * TweenService:GetValue(math.clamp(math.min(angularSpeed, math.abs(angleDiff/deltaTime) / angularSpeed), 0, 1), easingStyle, easingDirection) * angularSpeed * deltaTime
+end
+
+function CameraLook.new(instance: Instance)
+	local self = setmetatable({}, CameraLook)
+
+	self.Instance = instance
+	self._trove = Trove.new()
+
+	-- Wait for Humanoid to be added
+	local humanoid = instance:FindFirstChildWhichIsA("Humanoid")
+	while not humanoid do
+		humanoid = instance:FindFirstChildWhichIsA("Humanoid")
+		instance.ChildAdded:Wait()
+	end
+
+	-- Retrieve player root part
+	local rootPart = humanoid.RootPart
+
+	-- Speed for rotating the character itself
+	self.LookAngularSpeed = math.rad(400)
+
+	-- Disable automatic rotation
+	humanoid.AutoRotate = false
+
+	-- Speed for rotating the root
+	self.RootPitchSpeed = math.rad(80)
+	self.RootYawSpeed = math.rad(0)
+	-- Speed for rotating the waist
+	self.WaistPitchSpeed = math.rad(100)
+	self.WaistYawSpeed = math.rad(360)
+	-- Speed for rotating the neck
+	self.NeckPitchSpeed = math.rad(200)
+	self.NeckYawSpeed = math.rad(560)
+
+	-- Grab base C0 for root part (TODO: Clean this up)
+	local lowerTorso: BasePart = instance:WaitForChild("LowerTorso")
+	local root: Motor6D = lowerTorso and lowerTorso:WaitForChild("Root")
+	local rootC0 = root.C0
+
+	-- Reset joint angles
+	self._rootPitch = 0
+	self._rootYaw = 0
+	self._waistPitch = 0
+	self._waistYaw = 0
+	self._neckPitch = 0
+	self._neckYaw = 0
+
+	local player: Player? = Players:GetPlayerFromCharacter(instance)
+	local isLocalPlayer = player == Players.LocalPlayer
+	local mouse: Mouse? = if isLocalPlayer then player:GetMouse() else nil
+	self._trove:Connect(RunService.Stepped, function(_, deltaTime)
```

Comment:
Is there a reason we're using stepped over heartbeat?

Reply:
Yes, `Transform` must be applied on `Stepped` and cannot be applied at any other point. Additionally stepped is processed before physics which is what is desired here, as we want collisions with the character as well as things like bullet casts (likely to be done on Heartbeat since that'd be proper) to be visually consistent if performed client-sided.

Reply:
Cool